modules/gapi/src/backends/fluid/gfluidcore.cpp

```diff
@@ -117,8 +233,17 @@ static void run_addweighted(Buffer &dst, const View &src1, const View &src2,
     auto _beta  = static_cast<float>( beta  );
     auto _gamma = static_cast<float>( gamma );
 
-    for (int l=0; l < length; l++)
-        out[l] = addWeighted<DST>(in1[l], in2[l], _alpha, _beta, _gamma);
+    int x = 0;
+#if CV_SSE2 | CV_AVX2 | CV_AVX512_SKX
+    v_float32 a = vx_setall_f32(_alpha);
+    v_float32 b = vx_setall_f32(_beta);
+    v_float32 g = vx_setall_f32(_gamma);
+
+    x = addw_simd(in1, in2, out, a, b, g, length);
```

Comment:
Why?

I don't know any case where passing SIMD vectors through parameters (into non-inline function) can improve performance.

Reply:
I don't tried to improve performance with commit "Refactoring.Step2". I've just tried remove excess intermediate function which has the same name `addw_simd()` but look like this:
`template<typename SRC, typename DST>
CV_ALWAYS_INLINE int addw_simd(const SRC in1[], const SRC in2[], DST out[],
                               float _alpha, float _beta, float _gamma, int length)
{
    //Cases when dst type is float are successfully vectorized with compiler.
    if (std::is_same<DST, float>::value)
        return 0;

    v_float32 alpha = vx_setall_f32(_alpha);
    v_float32 beta = vx_setall_f32(_beta);
    v_float32 gamma = vx_setall_f32(_gamma);

    if (std::is_same<SRC, ushort>::value && std::is_same<DST, ushort>::value)
    {
        return addw_short2short(reinterpret_cast<const ushort*>(in1), reinterpret_cast<const ushort*>(in2),
                                reinterpret_cast<ushort*>(out), alpha, beta, gamma, length);
    }
    else if (std::is_same<SRC, short>::value && std::is_same<DST, short>::value)
    {
        return addw_short2short(reinterpret_cast<const short*>(in1), reinterpret_cast<const short*>(in2),
                                reinterpret_cast<short*>(out), alpha, beta, gamma, length);
    }
    else if (std::is_same<SRC, short>::value && std::is_same<DST, uchar>::value)
    {
        return addw_short2uchar(reinterpret_cast<const short*>(in1), reinterpret_cast<const short*>(in2),
                                reinterpret_cast<uchar*>(out), alpha, beta, gamma, length);
    }
    else if (std::is_same<SRC, ushort>::value && std::is_same<DST, uchar>::value)
    {
        return addw_short2uchar(reinterpret_cast<const ushort*>(in1), reinterpret_cast<const ushort*>(in2),
                                reinterpret_cast<uchar*>(out), alpha, beta, gamma, length);
    }
    else if (std::is_same<SRC, uchar>::value && std::is_same<DST, uchar>::value)
    {
        constexpr int nlanes = v_uint8::nlanes;

        if (length < nlanes)
            return 0;

        int x = 0;

        for (;;)
        {
            for (; x <= length - nlanes; x += nlanes)
            {
                v_float32 a1 = v_load_f32(reinterpret_cast<const uchar*>(&in1[x]));
                v_float32 a2 = v_load_f32(reinterpret_cast<const uchar*>(&in1[x + nlanes / 4]));
                v_float32 a3 = v_load_f32(reinterpret_cast<const uchar*>(&in1[x + nlanes / 2]));
                v_float32 a4 = v_load_f32(reinterpret_cast<const uchar*>(&in1[x + 3 * nlanes / 4]));
                v_float32 b1 = v_load_f32(reinterpret_cast<const uchar*>(&in2[x]));
                v_float32 b2 = v_load_f32(reinterpret_cast<const uchar*>(&in2[x + nlanes / 4]));
                v_float32 b3 = v_load_f32(reinterpret_cast<const uchar*>(&in2[x + nlanes / 2]));
                v_float32 b4 = v_load_f32(reinterpret_cast<const uchar*>(&in2[x + 3 * nlanes / 4]));

                v_int32 sum1 = v_round(v_fma(a1, alpha, v_fma(b1, beta, gamma))),
                        sum2 = v_round(v_fma(a2, alpha, v_fma(b2, beta, gamma))),
                        sum3 = v_round(v_fma(a3, alpha, v_fma(b3, beta, gamma))),
                        sum4 = v_round(v_fma(a4, alpha, v_fma(b4, beta, gamma)));

                vx_store(reinterpret_cast<uchar*>(&out[x]), v_pack_u(v_pack(sum1, sum2), v_pack(sum3, sum4)));
            }

            if (x < length)
            {
                x = length - nlanes;
                continue;  // process one more time (unaligned tail)
            }
            break;
        }
        return x;
    }
    return 0;
}`

Reply:
Functions `addw_short2uchar()`, `addw_short2short()` were renamed to `addw_simd()` and became templated with 2 template specializations. Please don't rely on github's diff. It often shows a difference incorrectly.
Since showed above intermediate function was removed, initialization of vectors alpha, beta, gamma had to be moved to function `run_addweighted()` on one call stack's level higher.

Reply:
So now `addw_simd()` look like this:

```
template<typename SRC, typename DST>
CV_ALWAYS_INLINE int addw_simd(const SRC in1[], const SRC in2[], DST out[],
                               const v_float32& alpha, const v_float32& beta,
                               const v_float32& gamma, int length)
{
    GAPI_Assert((std::is_same<DST, ushort>::value) ||
                (std::is_same<DST, short>::value));

    constexpr int nlanes = (std::is_same<DST, ushort>::value) ? static_cast<int>(v_uint16::nlanes) :
                                                                static_cast<int>(v_int16::nlanes);

    if (length < nlanes)
        return 0;

    int x = 0;
    for (;;)
    {
        for (; x <= length - nlanes; x += nlanes)
        {
            v_float32 a1 = v_load_f32(&in1[x]);
            v_float32 a2 = v_load_f32(&in1[x + nlanes / 2]);
            v_float32 b1 = v_load_f32(&in2[x]);
            v_float32 b2 = v_load_f32(&in2[x + nlanes / 2]);

            addw_short_store(&out[x], v_round(v_fma(a1, alpha, v_fma(b1, beta, gamma))),
                                      v_round(v_fma(a2, alpha, v_fma(b2, beta, gamma))));
        }

        if (x < length)
        {
            x = length - nlanes;
            continue;  // process one more time (unaligned tail)
        }
        break;
    }
    return x;
}

template<typename SRC>
CV_ALWAYS_INLINE int addw_simd(const SRC in1[], const SRC in2[], uchar out[],
                               const v_float32& alpha, const v_float32& beta,
                               const v_float32& gamma, int length)
{
    constexpr int nlanes = v_uint8::nlanes;

    if (length < nlanes)
        return 0;

    int x = 0;

    for (;;)
    {
        for (; x <= length - nlanes; x += nlanes)
        {
            v_float32 a1 = v_load_f32(&in1[x]);
            v_float32 a2 = v_load_f32(&in1[x + nlanes / 4]);
            v_float32 a3 = v_load_f32(&in1[x + nlanes / 2]);
            v_float32 a4 = v_load_f32(&in1[x + 3 * nlanes / 4]);
            v_float32 b1 = v_load_f32(&in2[x]);
            v_float32 b2 = v_load_f32(&in2[x + nlanes / 4]);
            v_float32 b3 = v_load_f32(&in2[x + nlanes / 2]);
            v_float32 b4 = v_load_f32(&in2[x + 3 * nlanes / 4]);

            v_int32 sum1 = v_round(v_fma(a1, alpha, v_fma(b1, beta, gamma))),
                    sum2 = v_round(v_fma(a2, alpha, v_fma(b2, beta, gamma))),
                    sum3 = v_round(v_fma(a3, alpha, v_fma(b3, beta, gamma))),
                    sum4 = v_round(v_fma(a4, alpha, v_fma(b4, beta, gamma)));

            vx_store(&out[x], v_pack_u(v_pack(sum1, sum2), v_pack(sum3, sum4)));
        }

        if (x < length)
        {
            x = length - nlanes;
            continue;  // process one more time (unaligned tail)
        }
        break;
    }
    return x;
}

template<typename SRC>
CV_ALWAYS_INLINE int addw_simd(const SRC*, const SRC*, float*,
                               const v_float32&, const v_float32&,
                               const v_float32&, int length)
{
    //Cases when dst type is float are successfully vectorized with compiler.
    return 0;
}
```

Reply:
But if you mind I can move 
```
v_float32 a = vx_setall_f32(_alpha);
v_float32 b = vx_setall_f32(_beta);
v_float32 g = vx_setall_f32(_gamma);
```
to each of the `addw_simd()` template specialization. But it cause code duplication.

Reply:
> `GAPI_Assert((std::is_same<DST, ushort>::value) ||
                (std::is_same<DST, short>::value));`

C++ way is using `std::enable_if` (SFINAE).

Reply:
> > `GAPI_Assert((std::is_same<DST, ushort>::value) ||                 (std::is_same<DST, short>::value));`
> 
> C++ way is using `std::enable_if` (SFINAE).

I'm confused a little. Could you please provide an example how I can apply `std::enable_if` for this case?

Reply:
I think it would be better and easier to add static assert.
@alalek could you please comment?

Reply:
OK, this may work here.

Reply:
Applied.

Reply:
> if you mind I can move

This should be moved into SIMD functions.

Think about runtime dispatching (not in a scope of this PR) where we don't even know SIMD size and registers types of underlying functions.
Which SIMD value do you want to send there?

---

Consider following these rules:
- no SIMD variables in dispatching code (generic C++ part)
- no SIMD arguments in SIMD functions which are called from generic C++ code.

Reply:
Applied.